### PR TITLE
Gamemode code cleanup.

### PR DESCRIPTION
**Pull request by DominikPanic**
Cleans up the gamemode and a bit of ticker code.

### CODE REVIEW

File: code/game/gamemodes/distress.dm [hunk 1]
```diff
@@ -198,57 +216,60 @@ datum/game_mode/proc/initialize_post_queen_list()
...
 	if(!mother.ovipositor || mother.is_mob_incapacitated(TRUE))
 		to_chat(xeno_candidate, "<span class='warning'>Mother is not in a state to receive us.</span>")
 		return FALSE
-	if(!xeno_bypass_timer && !isnewplayer(xeno_candidate))
+	if(!isnewplayer(xeno_candidate))
 		var/deathtime = world.time - xeno_candidate.timeofdeath
 		var/deathtimeminutes = round(deathtime / 600)
 		var/deathtimeseconds = round((deathtime - deathtimeminutes * 600) / 10,1)
-		if(deathtime < 3000 && ( !xeno_candidate.client.holder || !check_other_rights(xeno_candidate, R_ADMIN, FALSE)))
+		if(deathtime < 3000 && ( !xeno_candidate.client.holder || !check_rights(xeno_candidate, R_ADMIN)))
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]:
```suggestion
		if(deathtime < 5 MINUTES && ( !xeno_candidate.client.holder || !check_rights(xeno_candidate, R_ADMIN)))
```
This is in enough places to warrant its own define.

File: code/game/gamemodes/distress.dm [hunk 2]
```diff
@@ -272,19 +293,18 @@ datum/game_mode/proc/initialize_post_queen_list()
...
-			to_chat(xeno_candidate, "<span class='warning'>That player hasn't been away long enough. Please wait [300 - new_xeno.away_timer] second\s longer.</span>")
-			return FALSE
+	var/deathtime = world.time - xeno_candidate.timeofdeath
+	if(isnewplayer(xeno_candidate))
+		deathtime = 3000 //so new players don't have to wait to latejoin as xeno in the round's first 5 mins.
+	var/deathtimeminutes = round(deathtime / 600)
+	var/deathtimeseconds = round((deathtime - deathtimeminutes * 600) / 10,1)
+	if(deathtime < 3000)
+		to_chat(xeno_candidate, "<span class='warning'>You have been dead for [deathtimeminutes >= 1 ? "[deathtimeminutes] minute\s and " : ""][deathtimeseconds] second\s.</span>")
+		to_chat(xeno_candidate, "<span class='warning'>You must wait 5 minutes before rejoining the game!</span>")
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]: The define could auto-update this message.

File: code/game/gamemodes/distress.dm [hunk 3]
```diff
@@ -272,19 +293,18 @@ datum/game_mode/proc/initialize_post_queen_list()
...
+		return FALSE
+	if(new_xeno.away_timer < 300) //We do not want to occupy them if they've only been gone for a little bit.
+		to_chat(xeno_candidate, "<span class='warning'>That player hasn't been away long enough. Please wait [300 - new_xeno.away_timer] second\s longer.</span>")
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]: Another value that could be a define.

File: code/game/gamemodes/distress.dm [hunk 4]
```diff
@@ -754,89 +669,144 @@ datum/game_mode/proc/initialize_post_queen_list()
...
-	desc = "It looks way too dangerous to traverse. Best wait until it has cleared up."
-	icon = 'icons/effects/effects.dmi'
-	icon_state = "smoke"
-	opacity = TRUE
-
-/obj/effect/forcefield/fog/Initialize()
-	. = ..()
-	dir  = pick(CARDINAL_DIRS)
-	GLOB.fog_blockers += src
+	for(var/datum/mind/X in xenomorphs)
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]: This list should be clean, so no need to type check.

File: code/game/gamemodes/distress.dm [hunk 5]
```diff
@@ -1,194 +1,212 @@
...
+	var/num_humans = living_player_list[1]
+	var/num_xenos = living_player_list[2]
+
+	if(SSevacuation.dest_status == NUKE_EXPLOSION_FINISHED)
+		message_admins("Round finished: [MODE_GENERIC_DRAW_NUKE]")
+		round_finished = MODE_GENERIC_DRAW_NUKE
+	else if(!num_humans && num_xenos)
+		message_admins("Round finished: [MODE_INFESTATION_X_MAJOR]")
+		round_finished = MODE_INFESTATION_X_MAJOR
+	else if(num_humans && !num_xenos)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/game/gamemodes/distress.dm]: Out of scope

File: code/game/gamemodes/distress.dm [hunk 6]
```diff
@@ -1,194 +1,212 @@
...
+	for(var/i in possible_xenomorphs)
+		var/datum/mind/new_xeno = i
+		if(new_xeno.assigned_role || jobban_isbanned(new_xeno.current, ROLE_XENOMORPH) || is_banned_from(new_xeno.current?.ckey, ROLE_XENOMORPH))
+			continue
+		new_xeno.assigned_role = "Xenomorph"
+		xenomorphs += new_xeno
+		possible_xenomorphs -= new_xeno
 
-	return TRUE
+	if(length(xenomorphs) < xeno_starting_num)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/game/gamemodes/distress.dm]: Done

File: code/game/gamemodes/distress.dm [hunk 7]
```diff
@@ -1,194 +1,212 @@
...
-/datum/game_mode/proc/check_xeno_late_join(mob/xeno_candidate)
-	if(jobban_isbanned(xeno_candidate, ROLE_XENOMORPH) || is_banned_from(xeno_candidate.ckey, ROLE_XENOMORPH)) // User is jobbanned
-		to_chat(xeno_candidate, "<span class='warning'>You are banned from playing aliens and cannot spawn as a xenomorph.</span>")
-		return FALSE
+	for(var/i in possible_survivors)
+		var/datum/mind/new_survivor = i
+		if(new_survivor.assigned_role || jobban_isbanned(new_survivor.current, ROLE_SURVIVOR) || is_banned_from(new_survivor.current?.ckey, ROLE_SURVIVOR))
+			continue
+		new_survivor.assigned_role = "Survivor"
+		survivors += new_survivor
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/game/gamemodes/distress.dm]: Done

File: code/game/gamemodes/game_mode.dm [hunk 8]
```diff
@@ -124,159 +102,284 @@
...
+		qdel(i)
+		sleep(1)
+
+
+/datum/game_mode/proc/attempt_to_join_as_larva(mob/xeno_candidate)
+	to_chat(xeno_candidate, "<span class='warning'>This is not available in this gamemode.</span>")
+	return FALSE
+
+
+/datum/game_mode/proc/spawn_larva(mob/xeno_candidate, var/mob/living/carbon/Xenomorph/Queen/mother)
```
Comment (Rohesie) [inline code/game/gamemodes/game_mode.dm]: These procs should work in extended, for events.
Perhaps `/datum/game_mode/xenomorph/distress_signal` with a `/datum/game_mode/xenomorph/extended` option, and the same for a future WO.
Or just make these for all game modes.

File: code/game/gamemodes/huntergames.dm [hunk 9]
```diff
@@ -179,46 +179,20 @@ var/waiting_for_drop_votes = 0
...
 
-	if(!round_finished)
-		if(++round_checkwin >= 5) //Only check win conditions every 5 ticks.
-			check_win()
-			round_checkwin = 0
+/datum/game_mode/huntergames/proc/count_humans()
+	var/human_count = 0
 
-/datum/game_mode/huntergames/check_win()
+	for(var/mob/living/carbon/human/H in GLOB.alive_human_list)
```
Comment (Rohesie) [inline code/game/gamemodes/huntergames.dm]: No need to type check here nor under it.

File: code/game/objects/effects/decals/cleanable.dm [hunk 10]
```diff
@@ -12,4 +12,41 @@
...
+	loc_last_process = loc
+
+	
+/obj/effect/decal/cleanable/blood/splatter/animated/Destroy()
+	animation_destruction_fade(src)
+	STOP_PROCESSING(SSobj, src)
+	return ..()
+
+
+/obj/effect/decal/cleanable/blood/splatter/animated/process()
```
Comment (Rohesie) [inline code/game/objects/effects/decals/cleanable.dm]: Oh gods.

File: code/modules/admin/fun_verbs.dm [hunk 11]
```diff
@@ -781,24 +754,20 @@
 
 		dresscode = input("Select job equipment", "Select Equipment") as null|anything in sortList(job_outfits)
 		dresscode = job_outfits[dresscode]
-		if(isnull(dresscode))
-			return
 
-	if(dresscode == "-- Custom")
+	if(dresscode == "{Custom}")
```
Comment (Rohesie) [inline code/modules/admin/fun_verbs.dm]:
```suggestion
	else if(dresscode == "{Custom}")
```
Or alter the logic to make it a `switch`.

File: code/modules/admin/panels/mode_panel.dm [hunk 12]
```diff
@@ -79,16 +68,30 @@
...
-			if(!istype(H))
-				continue
-			dat += "<tr><td><a href='?priv_msg=[REF(H)]'>[H.real_name]</a>[H.client ? "" : " <i>(logged out)</i>"][H.stat == DEAD ? " <b><font color=red>(DEAD)</font></b>" : ""]</td>"
-			dat += "<td>[get_area(get_turf(H))]</td>"
-			dat += "<td><a href='?src=[ref];playerpanel=[REF(H)]'>PP</A></td></TR>"
-		dat += "</table>"
+	if(istype(SSticker.mode, /datum/game_mode/distress))
+		var/datum/game_mode/distress/D = SSticker.mode
+
+		if(length(D.xenomorphs))
```
Comment (Rohesie) [inline code/modules/admin/panels/mode_panel.dm]: Use `alive_xeno_list` instead?

File: code/modules/mob/dead/observer/observer.dm [hunk 13]
```diff
@@ -634,13 +613,17 @@
...
 
-	if(!client || !SSticker.mode.check_xeno_late_join(src))
+	if(!client)
 		return
 
 	if(!SSticker?.mode || SSticker.current_state < GAME_STATE_PLAYING)
 		to_chat(src, "<span class='warning'>The game hasn't started yet!</span>")
 		return
 
+	if(jobban_isbanned(src, ROLE_XENOMORPH) || is_banned_from(ckey, ROLE_XENOMORPH))
```
Comment (Rohesie) [inline code/modules/mob/dead/observer/observer.dm]: Might warrant a define.

File: code/modules/power/power.dm [hunk 14]
```diff
@@ -359,4 +359,90 @@
...
+		S.power_change()
+
+	for(var/obj/machinery/power/apc/C in GLOB.machines)
+		if(!C.cell || !is_mainship_level(C.z))
+			continue		
+		C.cell.charge = 0
+
+	playsound_z(3, 'sound/effects/powerloss.ogg')
+
+	sleep(100)
```
Comment (Rohesie) [inline code/modules/power/power.dm]: The shadow of CALLBACK prowls in the night!

File: code/modules/power/power.dm [hunk 15]
```diff
@@ -359,4 +359,90 @@
...
+		S.update_icon()
+		S.power_change()
+
+	for(var/obj/machinery/power/apc/C in GLOB.machines)
+		if(!C.cell || !is_mainship_level(C.z))
+			continue
+		C.cell.charge = C.cell.maxcharge
+
+
+	sleep(100)
```
Comment (Rohesie) [inline code/modules/power/power.dm]: An it lets no Killer sleep!

File: code/modules/power/power.dm [hunk 16]
```diff
@@ -359,4 +359,90 @@
...
+	for(var/obj/machinery/power/smes/S in GLOB.machines)
+		if(!is_mainship_level(S.z)) // Ship only
+			continue
+		S.charge = S.capacity
+		S.output_level = S.output_level_max
+		S.outputting = TRUE
+		S.update_icon()
+		S.power_change()
+
+	sleep(100)
```
Comment (Rohesie) [inline code/modules/power/power.dm]: J/K, I know it's out of scope and you hate CALLBACKs.

File: code/game/gamemodes/distress.dm [hunk 17]
```diff
@@ -0,0 +1,750 @@
...
+			command_announcement.Announce("An automated distress signal has been received from archaeology site \"Shiva's Snowball\", on border ice world \"Ifrit\". A response team from the [MAIN_SHIP_NAME] will be dispatched shortly to investigate.", "[MAIN_SHIP_NAME]")
+		if(MAP_BIG_RED)
+			command_announcement.Announce("We've lost contact with the Nanotrasen's research facility, [SSmapping.config.map_name]. The [MAIN_SHIP_NAME] has been dispatched to assist.", "[MAIN_SHIP_NAME]")
+		if(MAP_PRISON_STATION)
+			command_announcement.Announce("An automated distress signal has been received from maximum-security prison \"Fiorina Orbital Penitentiary\". A response team from the [MAIN_SHIP_NAME] will be dispatched shortly to investigate.", "[MAIN_SHIP_NAME]")
+
+
+/datum/game_mode/distress/process()
+	if(round_finished)
+		return FALSE
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/game/gamemodes/distress.dm]: Other modes may want to keep processing even after that

File: code/game/gamemodes/distress.dm [hunk 18]
```diff
@@ -0,0 +1,750 @@
...
+
+	return mother
+
+
+/datum/game_mode/distress/spawn_larva(mob/xeno_candidate, var/mob/living/carbon/Xenomorph/Queen/mother)
+	if(!xeno_candidate?.mind)
+		return FALSE
+
+	if(!SSticker.mode?.stored_larva || !istype(mother))
+		to_chat(xeno_candidate, "<span class='warning'>Something went awry. Can't spawn at the moment.</span>")
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/game/gamemodes/distress.dm]: Done

File: code/__DEFINES/mode.dm [hunk 19]
```diff
@@ -51,12 +46,61 @@
...
+
+#define SHUTTLE_TIME_LOCK 		10 MINUTES
+#define SHUTTLE_LOCK_COOLDOWN 	10 MINUTES
+#define SHUTTLE_LOCK_TIME_LOCK 	45 MINUTES
+
+#define COOLDOWN_COMM_REQUEST 	5 MINUTES
+#define COOLDOWN_COMM_MESSAGE 	1 MINUTES
+#define COOLDOWN_COMM_CENTRAL 	30 SECONDS
+
+#define XENO_AFK_TIMER			300
```
Comment (Rohesie) [inline code/__DEFINES/mode.dm]:
```suggestion
#define XENO_AFK_TIMER			300 //This is awful and could really be improved. Right now it's a number of life() ticks without a client.
```

File: code/game/gamemodes/distress.dm [hunk 20]
```diff
@@ -0,0 +1,691 @@
...
+	required_players = 2
+	flags_round_type = MODE_INFESTATION|MODE_FOG_ACTIVATED
+	flags_landmarks = MODE_LANDMARK_SPAWN_XENO_TUNNELS|MODE_LANDMARK_SPAWN_MAP_ITEM
+
+	var/list/survivors = list()
+
+	var/xeno_required_num 	  = 1
+	var/xeno_starting_num 	  = 0
+	var/surv_starting_num 	  = 0
+	var/merc_starting_num 	  = 0
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]: Legacy of halloween2016. Can we clean it?

File: code/game/gamemodes/distress.dm [hunk 21]
```diff
@@ -0,0 +1,691 @@
...
+						/obj/item/ammo_magazine/shotgun/buckshot = round(scale * 10),
+						/obj/item/ammo_magazine/shotgun/flechette = round(scale * 10),
+						/obj/item/cell/lasgun/M43 = round(scale * 25),
+
+						/obj/item/weapon/combat_knife = round(scale * 30),
+						/obj/item/weapon/throwing_knife = round(scale * 10),
+						/obj/item/storage/box/m94 = round(scale * 10),
+
+						/obj/item/attachable/flashlight = round(scale * 25),
+						/obj/item/attachable/bayonet = round(scale * 25),
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]:
```suggestion
						/obj/item/attachable/bayonet = round(scale * 25)
```

File: code/game/gamemodes/distress.dm [hunk 22]
```diff
@@ -0,0 +1,691 @@
...
+						/obj/item/explosive/grenade/cloakbomb = round(scale * 3),
+						/obj/item/storage/box/m94 = round(scale * 30),
+						/obj/item/device/flashlight/combat = round(scale * 5),
+						/obj/item/clothing/mask/gas = round(scale * 10)
+						)
+
+		CG.contraband = list(
+						/obj/item/weapon/gun/smg/ppsh = round(scale * 4),
+						/obj/item/weapon/gun/shotgun/double = round(scale * 2),
+						/obj/item/weapon/gun/shotgun/pump/bolt = round(scale * 2),
```
Comment (Rohesie) [inline code/game/gamemodes/distress.dm]:
```suggestion
						/obj/item/weapon/gun/shotgun/pump/bolt = round(scale * 2)
```

**Pull request closed without merging**